src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java

```diff
@@ -1536,6 +1536,17 @@ public void visitForeachLoop(JCEnhancedForLoop tree) {
                     elemtype = iterableParams.isEmpty()
                         ? syms.objectType
                         : types.wildUpperBound(iterableParams.head);
+
+                    // Check the return type of the method iterator()
+                    Symbol iterSymbol = rs.resolveInternalMethod(tree.pos(),
+                            loopEnv, exprType, names.iterator, List.nil(), List.nil());
+                    if (types.asSuper(iterSymbol.type.getReturnType(), syms.iteratorType.tsym) == null) {
+                        log.error(tree.pos(),
+                                Errors.OverrideIncompatibleRet(
```

Comment:
I believe here we don't need a new diagnostic - let's just reuse:

```
log.error(tree.expr.pos(),
                              Errors.ForeachNotApplicableToType(exprType,
                                                                Fragments.TypeReqArrayOrIterable));
```

On the basis that the provided type is NOT an well-formed `Iterable` type (because it has a bogus override).

The thing I'm a bit afraid, if we go down this path is that we will need to replicate all the override checks here - what if the method has the right signature but is private? It's still a bogus override.

I could be fine if the answer was "this is the bare minimum we need to verify to make sure code generation doesn't crash". But I'd like to see a comment on the code stating exactly that.

Do we have a similar problem with AutoCloseable? E.g. a close() method returning something odd?

Reply:
> I believe here we don't need a new diagnostic - let's just reuse:

Fixed.

> The thing I'm a bit afraid, if we go down this path is that we will need to replicate all the override checks here - what if the method has the right signature but is private? It's still a bogus override.
>
> I could be fine if the answer was "this is the bare minimum we need to verify to make sure code generation doesn't crash". But I'd like to see a comment on the code stating exactly that.

I added the comment to state it and added several test cases to test these situations.

> Do we have a similar problem with AutoCloseable? E.g. a close() method returning something odd?

The check of AutoCloseable is right. I tested the following cases locally. The compiler didn't crash and the error messages were reasonable.

```
public class T {
    public void test(String[] args) {
        try (Test test = new Test()) {
            int a = 1;
        } catch(Exception ex) {
            ex.printStackTrace();
        }
    }
}
class Test implements AutoCloseable {
    public int close() throws Exception {
        return 0;
    }
}
```

```
public class T {
    public void test(String[] args) {
        try (Test test = new Test()) {
            int a = 1;
        } catch(Exception ex) {
            ex.printStackTrace();
        }
    }
    class Test implements AutoCloseable {
        public int close() throws Exception {
            return 0;
        }
    }
}
```